src/honey_badger/honey_badger.rs

```diff
@@ -101,13 +105,19 @@ where
         self.has_input = true;
         let ser_prop =
             bincode::serialize(&proposal).map_err(|err| ErrorKind::ProposeBincode(*err))?;
-        let ciphertext = self
-            .netinfo
-            .public_key_set()
-            .public_key()
-            .encrypt_with_rng(&mut self.rng, ser_prop);
         let epoch = self.epoch;
-        let mut step = self.epoch_state_mut(epoch)?.propose(&ciphertext)?;
+        let require_decryption = self.encryption_schedule.use_on_epoch(epoch);
+        let prop = if require_decryption {
+            let ciphertext = self
+                .netinfo
+                .public_key_set()
+                .public_key()
+                .encrypt_with_rng(&mut self.rng, ser_prop);
+            bincode::serialize(&ciphertext).map_err(|err| ErrorKind::ProposeBincode(*err))?
+        } else {
+            ser_prop
+        };
```

Comment:
I'd prefer to move that part of the logic entirely into `EpochState`.
(Not your fault: It was already inconsistent in the first place to do the serialization and encryption in `HoneyBadger` and the deserialization and decryption in `EpochState`.)
Let's just input the `proposal` itself into `EpochState::propose`.

Reply:
This will require accepting or adding an RNG to `EpochState`. Thoughts?

Reply:
That's what I meant by 'accepting'. Does it matter if the RNG comes from `HoneyBadger`? IIRC, Rust doesn't have a good idiom for default/optional parameters yet. I'll probably end up doing this, it just feels like there should be a way to do it without passing in an RNG that is only sometimes used... It feels like we're going to end up with duplication somewhere no matter what though.

Reply:
Yes, it needs to come from `HoneyBadger`: We went to great lengths to make sure that we can configure all sources of randomness and have deterministic tests. Passing in a reference to the RNG is not expensive: It's no problem if it isn't always used.

Reply:
Hm, running into a borrow checker issue with:

```rust
let epoch_state = self.epoch_state_mut(epoch)?;
let rng = &mut self.rng;
epoch_state.propose(ser_prop, rng)?
```

I think I've got it sorted, but I'm not sure if there is a more idiomatic way to appease rust. Running tests then I'll push.

Reply:
I also might have an issue where clippy is telling me to do something but the compiler is unhappy when I do... Clippy might just be wrong on this one. I'm not sure if there is a way to tell it to ignore this specific case?

Reply:
I think the problem here is that `epoch_state_mut`, being a method on `self`, borrows _all_ of self and conflicts with the borrowed `rng`. Your workaround looks reasonable. (It's unfortunate that it's necessary, of course.)